src/Compiler/Microsoft.AspNetCore.Razor.Language/src/Syntax/SyntaxNode.cs

```diff
@@ -420,6 +422,207 @@ public bool IsEquivalentTo(SyntaxNode other)
+                if (tryWalkBackwards(stack, out foundToken))
+                {
+                    return foundToken;
+                }
+
+                // Encountered a newline while backtracking, so we need to walk forward instead.
+                return walkForward(stack);
+            }
+
+            bool tryWalkBackwards(PooledArrayBuilder<(SyntaxNode node, int nodeIndexInParent)> stack, [NotNullWhen(true)] out SyntaxToken? foundToken)
+            {
+                // Can't just pop the stack, we may need to rewalk from the start to find the next node if this fails
+                for (var originalStackPosition = stack.Count - 1; originalStackPosition >= 0; originalStackPosition--)
+                {
+                    var (node, nodeIndexInParent) = stack[originalStackPosition];
+
+                    switch (walkNodeChildren(node.Parent, nodeIndexInParent, walkBackwards: true, stopOnNewline: true, out foundToken))
+                    {
+                        case true:
+                            return true;
+                        case false:
+                            return false;
+                        case null:
+                            // Didn't find anything in this node, keep walking
+                            continue;
+                    }
+                }
+
+                // Walked all the way back to the end of the node that was requested and did not find either a newline or non-whitespace token. The user requested
+                // something out of range.
+                throw new ArgumentOutOfRangeException(nameof(position));
+            }
+
+            SyntaxToken walkForward(PooledArrayBuilder<(SyntaxNode node, int nodeIndexInParent)> stack)
+            {
+                while (stack.TryPop(out var entry))
+                {
+                    var (node, nodeIndexInParent) = entry;
+
+                    switch (walkNodeChildren(node.Parent, nodeIndexInParent, walkBackwards: false, stopOnNewline: false, out var foundToken))
+                    {
+                        case true:
+                            return foundToken;
+                        case null:
+                            // Didn't find anything in this node, keep walking
+                            continue;
+                        case false:
+                            // False is only returned when stopOnNewline is true
+                            return Assumed.Unreachable<SyntaxToken>();
+                    }
+                }
+
+                // Walked all the way forward to the end of the root that was requested and did not find any non-whitespace tokens. The user requested
+                // something out of range.
+                throw new ArgumentOutOfRangeException(nameof(position));
+            }
+
+            static bool? walkNodeChildren(SyntaxNode parent, int nodeIndexInParent, bool walkBackwards, bool stopOnNewline, [NotNullWhen(true)] out SyntaxToken? foundToken)
```

Comment:
Nit:

```suggestion
            static bool? walkNodeChildren(SyntaxNode parent, int startIndex, bool walkBackwards, bool stopOnNewline, [NotNullWhen(true)] out SyntaxToken? foundToken)
```

This is now being used as the startIndex, we don't really care where it came from.